impress-org/givewp-next-gen
# Feature: multi-step form design

## Description



This adds initial support to our donation form for multi-step style form designs.  Included here is an obvious named "Multi-Step Form Design" that enables the new `$isMultiStep` flag in its form design class and includes its own css.

From a high level, the donation form will take all the sections in the form builder and create a multi-step form using each section as a "step".

**Summary of the technical functionality**
- Form rendering loops through sections created in the form builder and creates new step objects based on the section details
- The current step renders its form/details.
- In order to proceed to next step it triggers (currently) front-end validation (later server-side validation as well)
- Once valid, the new step will be mounted/shown and the previous will be hidden
- _Note: the previous steps need to stay present in the DOM in order for fields like the gateways that need their scripts accessed when calling donate on the final step._
- The current step always holds the updated form values utilizing react-hook-form's defaultValues which is updated in a parent context provider.
- Whenever navigating to a previous step, all later steps are destroyed/unmounted to preserve integrity

## Affects



This is a _new_ form design that hooks into _new_ functionality that enables multi-step style forms.  Some existing functionality needed to be reused so some things were put into their own files.

## Visuals



<img width="576" alt="Screenshot 2023-05-02 at 1 51 16 PM" src="https://user-images.githubusercontent.com/10138447/235745288-cfec857c-6080-496a-85c2-fe5eae4063f4.png">

<img width="604" alt="Screenshot 2023-05-01 at 4 07 28 PM" src="https://user-images.githubusercontent.com/10138447/235522198-d1606c64-53f1-4a0e-b8dd-62bf2ea99a42.png">

<img width="575" alt="Screenshot 2023-05-01 at 4 07 46 PM" src="https://user-images.githubusercontent.com/10138447/235522218-b0c29a21-0efa-4f5e-a5cf-83f9b03979bf.png">


## Testing Instructions



- Pull branch
- Run `npm install && npm run dev`
- Create a new next gen form and select the new "Multi-Step" form design.
- Publish changes and use the form.
- Each section should correspond to a "step" in the multi-step form flow.

## Pre-review Checklist
